### ISSUE TITLE
As a User/Registered User, I want to Access an experiment referenced in EarthCODE Cat on CDSE, so that I can later execute it on CDSE

### ISSUE THREAD

**Issue by JanssenBrm** [ESA-EarthCODE/EarthCODE#10]

**Comment by JanssenBrm** [comment]
@GarinSmith - Could you confirm the following acceptance criteria for this ticket?

- [ ] A user can use a STAC URL, found for an experiment result in the EarthCODE catalogue, and use this reference to the product as the input for another experiment/workflow on openEO.

If you think of additional scenarios, please feel free to add them. I'd like to clarify this ticket, as the term "experiment result" suggests a final product. While such products can serve as inputs for subsequent workflows or experiments, they cannot be “executed” in the way implied by the ticket title. Additionally, what is expected with "Accessing" a result? What would be the user journey for discovering EarthCODE experiment results?

**Comment by JanssenBrm** [comment]
Comment from @GarinSmith 

@edobrowolska and @dgiomo,  I think this is a good point.  
We have three stories relating to accessing 1) Experiments, 2) Workflows and 3) Data/Datasets (Products)
To some degree these user stories are flexible depending upon the specific platform solutions and so we may need to adjust them for each platform.  Is this story mainly about the actual experiment and accessing the things defined in OGC API Records?
This will include the experiment metadata referencing the experiment itself,  so that we can later execute/reproduce  it.  See the experiment user story . https://github.com/ESA-EarthCODE/EarthCODE/issues/13

Maybe this user story should be ?
"As a User/Registered User, I want to Access experiment metadata in EarthCODE Cat originating from CDSE, so that I can later execute that experiment again on CDSE"
Done criteria :  A  User/Registered User can see the OGC API Records metadata that was published from CDSE.  

This could be a fairly simple story about access to the metadata and the reference to the experiment. Most of the effort here is for OSC.  It is different from the story about the Data/Datasets (Products) which we would want to access and view them on the platforms themselves.  Any thoughts?

Note that we probably care more about 1) Experiments and 3) Data/Datasets (Products) in the first instance.  
Note also the view of how this might map to openEO. E.g. something like:
Experiment = Process Graph 
Workflow =  User Defined Process 
Product = JSON Metada pointing at various file types

**Comment by JanssenBrm** [comment]
Thanks for your comment @GarinSmith  

However, if the user story changes to "As a User/Registered User, I want to Access experiment metadata in EarthCODE Cat originating from CDSE, so that I can later execute that experiment again on CDSE", it sounds like this is more a ticket for the EarthCODE cat instead of the platforms. The platform part seems to be covered by https://github.com/ESA-EarthCODE/EarthCODE/issues/9?

The question still remains on the scope of "accessing" a product. From an openEO perspective, this could mean taking the STAC URL to the product and using it in a new openEO process. The viewing part sounds more like something platform-independent. However, I'm happy to further discuss this in one of the upcoming meetings.

**Comment by JanssenBrm** [comment]
@GarinSmith  - I have updated the title of this ticket to refer to the actual experiment as the results are handled in https://github.com/ESA-EarthCODE/EarthCODE/issues/12.

**Comment by GarinSmith** [comment]
@JanssenBrm , I agree with this suggestion.  I also dicussed this with ESA this morning who were also in agreement. I will change the stories for the other platforms to be consistent.

Note that in the context of an experiment, access can mean access to metadata and access to the actual experiment as necessary.

**Comment by JanssenBrm** [comment]
Thank you @GarinSmith - is there anything expected from the platform regarding access to experiments? I can imagine that a user would go to the EarthCODE catalogue and search for the experiment. If the experiment is found by the user, they can execute it on one of the platforms. If this use case is correct, it looks like most of this scenario is already covered by existing tickets:

- Ensuring that the experiment is available in the EarthCODE catalogue (https://github.com/ESA-EarthCODE/EarthCODE/issues/9)
- Execute the experiment on the platform (https://github.com/ESA-EarthCODE/EarthCODE/issues/13)

In that case, it is unclear to us what the scope/acceptance criteria of this ticket would be for platforms. However, I'm probably missing a crucial part of the use case, so feel free to share any thoughts on this.

**Comment by GarinSmith** [comment]
@JanssenBrm
I think the first user story above relates to access to the experiment metadata appearing in the catalogue.
The second user story above relates to execution of the experiment  
This user story is about access to the experiment referenced by the metadata in the catalogue.
This is important as an intermediate step and also to allow someone to simply look at an experiment without running it.

I suggest the following done criteria. Done criteria:  
"As a User/Registered User, I can Access the experiment referenced by the EarthCODE Catalogue experiment metadata"
"This experiment might include code and/or configuration"
"This experiment might be stored in GitHub and/or on the CDSE platform"

Does that make sense?

**Comment by JanssenBrm** [comment]
@GarinSmith - Thank you very much for the added feedback. It is becoming clearer to me that there is an intermediate step between publishing and executing an experiment. It could be useful to view the details of an experiment on the platforms themselves as well.

Translating this openEO specifically, where an openEO process graph defines an experiment, the criteria could be:

- As a User/Registered User, I can view an experiment (code and configuration) referenced by the EarthCODE Catalogue within the openEO tools on CDSE.
- This experiment might be stored in GitHub and/or on CDSE

What do you think?

**Comment by GarinSmith** [comment]
@JanssenBrm, 

Yes, I agree. However, I was expecting that the capability for the step (view experiment on the platform) would have been demonstrated as part of the existing Develop epic, when the experiment was originally developed using the openEO tools on CDSE.  I am assuming platforms will very likely already have a way to load an existing experiment as part of the current solution. 

In the first instance, we will show how an experiment can be reproduced on the originating (CDSE) platform to keep things simple.  Given that the experiment might be stored in GitHub and/or on the CDSE platform, then CDSE is free to choose where it wants to load an experiment from.

So hopefully we can re-use the approach in the development Epic, but yes Execute assumes (probably) that we will first view experiment on the platform to enable execution. 

Note that in the technical architecture document on page 65 we have some RESTful services
Develop Experiment - CDSE - RESTful - Develop based upon a unique an experiment id. VITO to link/update

This RESTful API is designed to be flexible,  but I was imagining that Develop Experiment could reference and then load and "view the details of an experiment on the platform", so that the experiment could be executed and reproduced using the existing Development process. 

Once we have this working and reproducing on the originating platform, we could make it more complicated and perhaps look to reproduce an experiment on a new platform. We may also want to consider the use of OGC API Processes to enable a polymorphic approach to reproducibility across more than one platform. See also
Execute Experiment - CDSE/CREODIAS - RESTful - Execute an experiment with a unique id. VITO to provide link/update

In the first instance the approach is designed to be flexible,  so that it can help platforms demonstrate reproducibility in a way that works for them.  

**Perhaps the Access Epic could just implement the Develop Experiment RESTful API to view an existing experiment?** 

The Execute Epic only then needs to repeat the existing execute capability. Hopefully this is just a button press after we "view the details of an experiment on the platform"

Perhaps it would help to have a diagram to reinforce the data flow/process?

Sorry for the delay in responding, I just spotted this.

**Comment by JanssenBrm** [comment]
We have identified the following important considerations:

- To implement access to an experiment, the issue at https://github.com/ESA-EarthCODE/EarthCODE/issues/7 must be resolved first. This includes storing not only the metadata in the EarthCODE catalogue but also all essential assets, such as the openEO process graph and output products, in permanent storage.
- Accessing the experiment requires a reference to the openEO process graph (see previous point). If this is publicly available, you can use existing tools to access and execute the experiment:
  - Load the experiment using the `datacube_from_json` process via the openEO Python Client (https://open-eo.github.io/openeo-python-client/datacube_construction.html#some-examples).
  - The openEO Web Editor allows you to import a workflow from an external source, as illustrated in the screenshot below.
  
<img width="1278" alt="Image" src="https://github.com/user-attachments/assets/973b5f7d-f72c-4a61-8290-b2d48efe3952" />

@GarinSmith  - Based on the precondition stated in the first point, it appears that CDSE may already cover the access and execution of experiments. What are your thoughts? Please feel free to add specific user requirements or examples if you notice anything missing.

**Comment by GarinSmith** [comment]
@JanssenBrm I think issues 13 and 14 are resolved.  
However I think we need a user journey to allow 
"As a User/Registered User, I can Access the experiment referenced by the EarthCODE Catalogue experiment metadata"

I think EarthCODE Catalogue needs a reference to the CDSE openEO platform for a specific experiment id. 
E..g   https://dataspace.copernicus.eu?experimentid=1234567 
@Schpidi does that make sense following our previous discussions on this topic?

@JanssenBrm 
I think we need to add a story to show how Vito can add an experiment to the PRR.  
We discussed this last week.  I believe you thought it would be straightforward to use 
the script provided by the PRR to move your product from its current temporary S3 bucket into the PRR?
